sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.Azure.Functions.Worker.Sdk.Generators
+{
+    /// <summary>
+    /// Generates a class that implements IFunctionMetadataProvider and the method GetFunctionsMetadataAsync() which returns a list of IFunctionMetadata. 
+    /// This source generator indexes a Function App and explicitly creates a list of DefaultFunctionMetadata (which implements IFunctionMetadata) from the functions defined
+    /// in the user's compilation. This allows the worker to index functions at build time, rather than waiting for the process to start.
+    /// </summary>
+    [Generator]
+    public partial class FunctionMetadataProviderGenerator : ISourceGenerator
```

Comment:
@kshyju has PR #1006 up to convert from `ISourceGenerator` to `IIncrementalGenerator`. Is it low cost to do similarly in this PR?

Reply:
I was planning to add this later in another PR - @kshyju has mentioned to me that upgrading to use `IIncrementalGenerator` introduced some issues with dependencies and testing in the repo that he's working on fixing, so I'm going to hold off for now.